File: scala/src/main/scala/problems/Problem004FirstNonRepeatedChar.scala
package problems
object Problem004FirstNonRepeatedChar {
  def firstNonRepeated(s:String): Option[Char] = {
    if (s==null) None
    else {
      val counts = s.foldLeft(scala.collection.mutable.LinkedHashMap.empty[Char,Int])((m,c) => { m(c)=m.getOrElse(c,0)+1; m })
      counts.collectFirst{ case (ch,c) if c==1 => ch }
    }
  }
}

File: java/src/main/java/problems/Problem004FirstNonRepeatedChar.java
package problems;


import java.util.*;
public class Problem004FirstNonRepeatedChar {
    public static Character firstNonRepeated(String s) {
        if(s==null) return null;
        Map<Character,Integer> c=new LinkedHashMap<>();
        for(char ch: s.toCharArray()) c.put(ch, c.getOrDefault(ch,0)+1);
        for(var e: c.entrySet()) if(e.getValue()==1) return e.getKey();
        return null;
    }
}

